src/diag/map_editor.cpp

```diff
@@ -5,149 +5,96 @@
-        case MAP_ID_A_DIESEL_DOWNSHIFT:
-            return A_DIESEL_DOWNSHIFT_MAP;
-        case MAP_ID_A_PETROL_UPSHIFT:
-            return A_PETROL_UPSHIFT_MAP;
-        case MAP_ID_A_PETROL_DOWNSHIFT:
-            return A_PETROL_DOWNSHIFT_MAP;
-        default:
-            return nullptr;
+#define CHECK_MAP(map_id) \
+    StoredTcuMap* ptr = get_map(map_id); \
+    if (ptr == nullptr) { \
+        return NRC_SUB_FUNC_NOT_SUPPORTED_INVALID_FORMAT; \
     }
-}
 
-uint8_t MapEditor::read_map_data(uint8_t map_id, uint16_t *dest_size, int16_t** buffer) {
-    uint8_t map_idx = map_id & 0b1111111;
-    bool is_default_map = (map_id & 0b10000000) != 0;
-    if (map_idx >= 0x01 && map_idx <= 0x0C) {
-        // Shift map
-        int16_t* b = (int16_t*)heap_caps_malloc(SHIFT_MAP_SIZE*sizeof(int16_t), MALLOC_CAP_SPIRAM);
-        if (b == nullptr) {
-            ESP_LOGE("MAP_EDITOR_R", "Could not allocate read array!");
-            return NRC_GENERAL_REJECT;
-        }
-        if (is_default_map) {
-            // Read default map from flash
-            const int16_t* default_map = get_default_map(map_idx);
-            if (default_map == nullptr) {
-                ESP_LOGE("MAP_EDITOR_R", "default map is invalid!?");
-                delete b;
-                return NRC_GENERAL_REJECT;
-            }
-            memcpy(b, default_map, SHIFT_MAP_SIZE*sizeof(int16_t));
-            *buffer = b;
-            *dest_size = SHIFT_MAP_SIZE*sizeof(int16_t);
-            return 0;
-        } else {
-            const char* name = map_id_to_name(map_idx);
-            if (name == nullptr) {
-                ESP_LOGE("MAP_EDITOR_R", "map name is null!?");
-                delete b;
-                return NRC_GENERAL_REJECT;
-            }
-            if (EEPROM::read_nvs_map_data(name, b, nullptr, SHIFT_MAP_SIZE)) {
-                // OK, copy the pointer
-                *buffer = b;
-                *dest_size = SHIFT_MAP_SIZE*sizeof(int16_t);
-                return 0;
-            } else {
-                // De-allocate
-                ESP_LOGE("MAP_EDITOR_R", "read_nvs_map_data failed!");
-                delete b;
-                return NRC_GENERAL_REJECT;
-            }
-        }
+uint8_t MapEditor::read_map_data(uint8_t map_id, bool is_default, uint16_t *dest_size, int16_t** buffer) {
+    CHECK_MAP(map_id)
+    // Map valid
+    int size = ptr->get_map_element_count();
+    int16_t* b = (int16_t*)heap_caps_malloc(size*sizeof(int16_t), MALLOC_CAP_SPIRAM);
+    if (b == nullptr) {
+        ESP_LOGE("MAP_EDITOR_R", "Could not allocate read array!");
+        return NRC_GENERAL_REJECT;
+    }
+    const int16_t* data;
+    if (is_default) {
+        data = ptr->get_default_map_data();
     } else {
-        return NRC_SUB_FUNC_NOT_SUPPORTED_INVALID_FORMAT;
+        data = (const int16_t*)ptr->get_current_map_data();
     }
+    memcpy(b, data, size*sizeof(int16_t));
+    *buffer = b;
+    *dest_size = size*sizeof(int16_t);
+    return 0;
```

Comment:
Is there an equivalent to "NRC_GENERAL_REJECT", like "NRC_GENERAL_OK"?

Reply:
I was trying to keep to KWP2000 naming convension, there is no such thing as "Negative response code 'OK'", so 0 was chosen where no negative value was needed

Reply:
Haha, great. A "magic number", if everything is fine. :')